git.go

```diff
@@ -113,3 +122,55 @@ func CloneRepo(ctx context.Context, opts CloneRepoOptions) (bool, error) {
 	}
 	return true, nil
 }
+
+func ReadPrivateKey(path string) (gossh.Signer, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("open private key file: %w", err)
+	}
+	var buf bytes.Buffer
+	if _, err := io.Copy(&buf, f); err != nil {
+		return nil, fmt.Errorf("read private key file: %w", err)
+	}
+	k, err := gossh.ParsePrivateKey(buf.Bytes())
+	if err != nil {
+		return nil, fmt.Errorf("parse private key file: %w", err)
+	}
+	return k, nil
+}
+
+// GenerateKnownHosts dials the server located at gitURL and fetches the SSH
+// public keys returned in a format accepted by known_hosts.
+func GenerateKnownHosts(log LoggerFunc, gitURL *url.URL) ([]byte, error) {
+	var buf bytes.Buffer
+	conf := &gossh.ClientConfig{
+		// Accept and record all host keys
+		HostKeyCallback: func(dialAddr string, addr net.Addr, key gossh.PublicKey) error {
+			h := strings.Split(dialAddr, ":")[0]
+			k64 := base64.StdEncoding.EncodeToString(key.Marshal())
+			log(codersdk.LogLevelInfo, "ssh keyscan: %s %s %s", h, key.Type(), k64)
+			buf.WriteString(fmt.Sprintf("%s %s %s\n", h, key.Type(), k64))
+			return nil
+		},
+	}
+	dialAddr := hostPort(gitURL)
+	client, err := gossh.Dial("tcp", dialAddr, conf)
+	if err != nil {
+		// The dial may fail due to no authentication methods, but this is fine.
+		if netErr, ok := err.(net.Error); ok {
+			return nil, fmt.Errorf("keyscan %s: %w", dialAddr, netErr)
+		}
+		// If it's not a net.Error then we will assume we were successful.
+	} else {
+		_ = client.Close()
+	}
+	return buf.Bytes(), nil
+}
+
+func hostPort(u *url.URL) string {
+	p := 22 // assume default SSH port
+	if _p, err := strconv.Atoi(u.Port()); err == nil {
```

Comment:
Nit: Prefer `pp`, underscores are unusual in variable names.

Reply:
You are using too much TS Cian 😆